### PR TITLE
added task frogger game

### PR DESCRIPTION
**Pull request by chris-voitova**
### Frogger Game

[Demo](https://chris-voitova.github.io/frontend-nanodegree-arcade-game/) |
[Code base](https://github.com/chris-voitova/frontend-nanodegree-arcade-game)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/chris-voitova/frogger-game/app.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
+const field = {
+  width: 505,
```
Comment (nazmariam) [inline submissions/chris-voitova/frogger-game/app.js:2]: It's better to move such 'magic numbers' (like 505, 606, 400, 450, 50) into constants with obvious names

File: submissions/chris-voitova/frogger-game/app.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+// Draw the enemy on the screen, required method for game
+Enemy.prototype.render = function () {
+  ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+// Now write your own player class
+// This class requires an update(), render() and
+// a handleInput() method.
+
+const Player = function (x, y, step = 50, width = 101, height = 101) {
```
Comment (nazmariam) [inline submissions/chris-voitova/frogger-game/app.js:47]: Same here

File: submissions/chris-voitova/frogger-game/app.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    this.goToInitPosition();
+  }
+};
+
+Player.prototype.render = function () {
+  ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+Player.prototype.handleInput = function (key) {
+  if (key === "up") {
```
Comment (nazmariam) [inline submissions/chris-voitova/frogger-game/app.js:92]: Mb we can use switch case here? 🤔

Comment (chris-voitova) [reply submissions/chris-voitova/frogger-game/app.js:92]: I thought about that, but something went wrong 😆

File: submissions/chris-voitova/frogger-game/app.js (hunk 1 above)
Comment (chris-voitova) [reply submissions/chris-voitova/frogger-game/app.js]: fixed

File: submissions/chris-voitova/frogger-game/app.js (hunk 2 above)
Comment (chris-voitova) [reply submissions/chris-voitova/frogger-game/app.js]: also fixed

File: submissions/chris-voitova/frogger-game/app.js (hunk 3 above)
Comment (chris-voitova) [reply submissions/chris-voitova/frogger-game/app.js]: rewrited to switch case

**Comment by nazmariam** [comment]
Looks wonderful!
50 cent: as for me, it's not obvious for player that collision zone is so big, mb it's better limit the perimeters of conflicts to the size of the player and the enemy?

File: submissions/chris-voitova/frogger-game/app.js (hunk 2 above)
Comment (OleksiyRudenko) [reply submissions/chris-voitova/frogger-game/app.js]: The following is not a code review and not a requirement to fix anything. Consider it a source of ideas you may consider to use in the future.

You might have grouped constants like e.g.
```
// related constants in a common declaration
const playerWidth = 101,
  playerHeight = 101,
  playerFieldStep = 50;

// ... or related constants grouped in an object literal; 
// this also makes comment like "// player" unnecessary as the const name is pretty descriptive per se
const playerProps = {
  step: 50,
  width: 101,
  height: 101
};
```

Why grouping? This way you lessen the chances that someone else would insert an unrelated `const` declaration between related constants.

The above is true for `let` declarations as well.

You can also use variables and constants as default values in function declarations.
E.g.
```
const Player = function(x, y, width = playerProps.width) { 
// ...
```

P.S. I didn't check other changes.

**Review by babayK0 — CHANGES_REQUESTED** [review]
Nice work! 
Lets make some improvements in your code :smile_cat: 

First of all it seems like i won  without reaching the edge. Fix this, please. 

Screenshot from 2020-10-16 00-29-38

Also see my comments below)

File: submissions/chris-voitova/frogger-game/app.js [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+// Now write your own player class
+// This class requires an update(), render() and
+// a handleInput() method.
+
+const Player = function (positionX, positionY, width, height, stepOnField) {
+  this.sprite = "images/char-boy.png";
+  this.positionX = positionX;
+  this.positionY = positionY;
+  this.width = width;
+  this.height = height;
```
[the remark this answers is not in the document]
Comment (chris-voitova) [reply submissions/chris-voitova/frogger-game/app.js]: thanks, made it

File: submissions/chris-voitova/frogger-game/app.js [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+// Draw the enemy on the screen, required method for game
+Enemy.prototype.render = function () {
+  ctx.drawImage(Resources.get(this.sprite), this.positionX, this.positionY);
+};
+
+// Now write your own player class
+// This class requires an update(), render() and
+// a handleInput() method.
+
+const Player = function (positionX, positionY, width, height, stepOnField) {
```
[the remark this answers is not in the document]
Comment (chris-voitova) [reply submissions/chris-voitova/frogger-game/app.js]: U're right. Cleaned it

File: submissions/chris-voitova/frogger-game/app.js [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+};
+
+// Now instantiate your objects.
+
+const enemy1 = new Enemy(0, 280, 60, enemyWidth, enemyHeight);
+const enemy2 = new Enemy(0, 200, 100, enemyWidth, enemyHeight);
+const enemy3 = new Enemy(30, 130, 80, enemyWidth, enemyHeight);
+
+// Place all enemy objects in an array called allEnemies
+const allEnemies = [enemy1, enemy2, enemy3];
```
[the remark this answers is not in the document]
Comment (chris-voitova) [reply submissions/chris-voitova/frogger-game/app.js]: Great advice! Implemented in this way. It's ok?
`const enemiesValues = [
  [0, 280, 60],
  [0, 200, 100],
  [30, 130, 80],
];`
`const allEnemies = enemiesValues.map((value) => new Enemy(...value));`

File: submissions/chris-voitova/frogger-game/app.js [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+  });
+};
+
+Player.prototype.goToInitPosition = function () {
+  this.positionX = this.initPosition.positionX;
+  this.positionY = this.initPosition.positionY;
+};
+
+Player.prototype.update = function () {
+  this.checkCollisions(allEnemies);
```
[the remark this answers is not in the document]
Comment (chris-voitova) [reply submissions/chris-voitova/frogger-game/app.js]: Thank! I rewrote and was able to called checkCollisions in update of engine.js

**Comment by chris-voitova** [comment]
> Nice work!
> Lets make some improvements in your code 😸
> 
> First of all it seems like i won without reaching the edge. Fix this, please.
> 
> Screenshot from 2020-10-16 00-29-38
> 
> Also see my comments below)

fixed

**Comment by chris-voitova** [comment]
> Looks wonderful!
> 50 cent: as for me, it's not obvious for player that collision zone is so big, mb it's better limit the perimeters of conflicts to the size of the player and the enemy?

I made some updates and now zone became a bit smaller

File: submissions/chris-voitova/frogger-game/app.js (hunk 2 above)
Comment (chris-voitova) [reply submissions/chris-voitova/frogger-game/app.js]: Thanks a lot! I understood your ideas, and thought about them. Finally, I reorganised my code for a better experience.

**Comment by babayK0** [comment]
> > Nice work!
> > Lets make some improvements in your code smile_cat
> > First of all it seems like i won without reaching the edge. Fix this, please.
> > Screenshot from 2020-10-16 00-29-38
> > Also see my comments below)
> 
> fixed


Something goes wrong with your collisions check. I am on edge but Enemy can touch me and i am lose. Also i need do one more step when i am on the edge to win game.


Screenshot from 2020-10-22 16-35-31

File: submissions/chris-voitova/frogger-game/app.js [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+      alert("you are lose!");
+      this.goToInitPosition();
+    }
+  }
+  goToInitPosition() {
+    this.positionX = this.initPosition.positionX;
+    this.positionY = this.initPosition.positionY;
+  }
+  update() {
+    if (this.positionY <= field.waterCoord) {
```
Comment (babayK0) [inline submissions/chris-voitova/frogger-game/app.js:100]: Move this check into `checkCollisions`. `update` must do update of your player only.

**Review by babayK0 — CHANGES_REQUESTED** [review]

File: submissions/chris-voitova/frogger-game/app.js [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+// Place the player object in a variable called player
+
+const player = new Player(215, 450);
+
+function checkCollisions() {
+  allEnemies.forEach((enemy) => {
+    player.checkCollisions(enemy);
+  });
+}
+
```
[the remark this answers is not in the document]
Comment (chris-voitova) [reply submissions/chris-voitova/frogger-game/app.js:153]: done image

File: submissions/chris-voitova/frogger-game/app.js (hunk 8 above)
Comment (chris-voitova) [reply submissions/chris-voitova/frogger-game/app.js:100]: moved this to checkWin method, because checkCollision called "number of enemies" times

**Comment by chris-voitova** [comment]
@babayK0 all items fixed, please check

File: submissions/chris-voitova/frogger-game/app.js [hunk 10]
```diff
@@ -0,0 +1,165 @@
+const field = {
```
[the remark this answers is not in the document]
Comment (babayK0) [reply submissions/chris-voitova/frogger-game/app.js:1]: @chris-voitova watch this please.
It is not a blocker, if you want you can improve your code.

**Review by babayK0 — APPROVED** [review]
Good job ! :+1:

**Pull request merged**